### PR TITLE
Find path and line using original sourcemap

### PR DESCRIPTION
**Pull request by iunanua**
### What does this PR do?

Exports a new method called `getOriginalPathAndLine` in charge of read from disk a javascript file, check if it contains a sourceMapping, parse or load it and search the corresponding token to the line and column passed as arguments.
If line and column are found it will return an object with the path, line and column in the original source file.

### Motivation
When IAST detects a vulnerability, it contains the location (file and line) where the vulnerability has been found. If the file points to a source file, like in the case of a transpiled typescript or minimized js files, we can resolve the vulnerability location to the original source file.

### Checklist

- [ ] The CHANGELOG.md has been updated
- [x] Unit tests have been updated and pass
- [ ] If known, an appropriate milestone has been selected

### CODE REVIEW

File: js/source-map/index.js [hunk 1]
```diff
@@ -54,8 +57,30 @@ function getSourcePathAndLineFromSourceMaps (filename, line, column = 0) {
...
 
+function getSourcePathAndLineFromSourceMaps (filename, line, column = 0) {
+  const sourceMap = rewrittenSourceMapsCache.get(filename)
+  return getPathAndLine(sourceMap, filename, line, column)
+}
+
+function getOriginalPathAndLine (filename, line, column = 0) {
+  if (filename && line) {
+    try {
+      let sourceMap = originalSourceMapsCache.get(filename)
```
Comment (iunanua) [inline js/source-map/index.js:68]: We could filter some files checking if the file has been rewritten before: only if `filename` is in `rewrittenSourceMapsCache` continue with the rest of the logic. That will avoid some vulnerability resolution in third-party libs.

File: test/stack_trace.spec.js [hunk 2]
```diff
@@ -60,18 +60,17 @@ describe('V8 prepareStackTrace', () => {
...
-    }
+const nodeSourceMap = require('../js/source-map/node_source_map')
+const { expect } = require('chai')
+const readFileSync = function (filename) {
+  if (filename.indexOf('.map') > 0) {
+    return fs.readFileSync(path.join(sourceMapResourcesPath, path.basename(filename)))
+  } else if (filename.indexOf('.js') > 0) {
+    return fs.readFileSync(path.join(sourceMapResourcesPath, path.basename(filename)))
   }
+}
```
Comment (uurien) [inline test/stack_trace.spec.js:71]:
```suggestion
  if (filename.indexOf('.map') > 0 || filename.indexOf('.js') > 0) {
    return fs.readFileSync(path.join(sourceMapResourcesPath, path.basename(filename)))
  }
```

File: test/stack_trace.spec.js [hunk 3]
```diff
@@ -140,3 +139,81 @@ describe('getFilenameFromSourceMap', () => {
...
+    './node_source_map': nodeSourceMap,
+    fs: {
+      readFileSync: (filename) => {
+        switch (filename) {
+          case 'no-sourcemap':
+            return ''
+          case 'error':
+            throw new Error('errr')
+          default:
+            return readFileSync(filename).toString()
```
Comment (uurien) [inline test/stack_trace.spec.js:170]: why are you doing `.toString` with the result of the original `readFileSync` to create the mock of `readFileSync`? I think that this should be:
```suggestion
            return readFileSync(filename)
```

File: test/resources/stacktrace-sourcemap/test-min.min.js [hunk 4]
```diff
@@ -1,5 +1,4 @@
-function test() {
+export function test() {
```
Comment (uurien) [inline test/resources/stacktrace-sourcemap/test-min.min.js:1]: You have updated this file but not the `test-min.min.js.map` file

Comment (iunanua) [reply test/resources/stacktrace-sourcemap/test-min.min.js:1]: It was a try to minimize the file but `prettier` is automatically formatting it after commiting or pushing it :S

File: js/source-map/index.js [hunk 5]
```diff
@@ -54,8 +57,31 @@ function getSourcePathAndLineFromSourceMaps (filename, line, column = 0) {
...
+      if (sourceMap === undefined) {
+        const filePath = getFilePathFromName(filename)
+        sourceMap = generateSourceMapFromFileContent(fs.readFileSync(filename).toString(), filePath)
+        originalSourceMapsCache.set(filename, sourceMap || null)
+      }
+      return getPathAndLine(sourceMap, filename, line, column)
+    } catch (e) {
+      // do nothing
```
[the remark this answers is not in the document]
Comment (iunanua) [reply js/source-map/index.js]: done!

**Review by uurien — APPROVED** [review]

**Pull request merged**